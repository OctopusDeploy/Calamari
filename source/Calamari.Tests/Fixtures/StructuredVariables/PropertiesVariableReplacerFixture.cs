using System;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;
using Assent;
using Calamari.Common.Features.StructuredVariables;
using Calamari.Common.Plumbing.Variables;
using Calamari.Tests.Helpers;
using FluentAssertions;
using NUnit.Framework;

namespace Calamari.Tests.Fixtures.StructuredVariables
{
    [TestFixture]
    public class PropertiesVariableReplacerFixture : VariableReplacerFixture
    {
        public PropertiesVariableReplacerFixture() : base((fs, log) => new PropertiesFormatVariableReplacer(fs, log))
        {
        }

        [Test]
        public void DoesNothingIfThereAreNoVariables()
        {
            var vars = new CalamariVariables();
            RunTest(vars, "example.properties");
        }

        [Test]
        public void DoesNothingIfThereAreNoMatchingVariables()
        {
            Replace(new CalamariVariables { { "Non-matching", "variable" } }, "example.properties");

            Log.MessagesInfoFormatted.Should().Contain(StructuredConfigMessages.NoStructuresFound);
            Log.MessagesVerboseFormatted.Should().NotContain(m => Regex.IsMatch(m, StructuredConfigMessages.StructureFound(".*")));
        }

        [Test]
        public void HandlesAnEmptyFile()
        {
            var vars = new CalamariVariables();
            RunTest(vars, "blank.properties");
        }

        [Test]
        public void CanReplaceASimpleKeyValuePair()
        {
            var vars = new CalamariVariables
            {
                { "key1", "new-value" }
            };
            RunTest(vars, "example.properties");

            Log.MessagesVerboseFormatted.Count(m => Regex.IsMatch(m, StructuredConfigMessages.StructureFound(".*"))).Should().Be(1);
            Log.MessagesVerboseFormatted.Should().Contain(StructuredConfigMessages.StructureFound("key1"));
            Log.MessagesInfoFormatted.Should().NotContain(StructuredConfigMessages.NoStructuresFound);
        }

        [Test]
        public void CanReplaceAKeyThatContainsAPhysicalNewLine()
        {
            var vars = new CalamariVariables
            {
                { "key3", "new-value" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void OnlyReplacesExistingKeys()
        {
            var vars = new CalamariVariables
            {
                { "no-such-key", "new-value" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void EscapesSlashesInValues()
        {
            var vars = new CalamariVariables
            {
                { "key1", "new\\value" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void EscapesDosNewLinesInValues()
        {
            var vars = new CalamariVariables
            {
                { "key1", "new\r\nvalue" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void EscapesNixNewLinesInValues()
        {
            var vars = new CalamariVariables
            {
                { "key1", "new\nvalue" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void EscapesTabsInValues()
        {
            var vars = new CalamariVariables
            {
                { "key1", "new\tvalue" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void EscapesALeadingSpaceInValues()
        {
            var vars = new CalamariVariables
            {
                { "key1", "  new value" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void CanReplaceAValueInAKeyThatDoesntHaveAValue()
        {
            var vars = new CalamariVariables
            {
                { "key5", "  new value" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void CanReplaceAValueInAKeyThatHasNeitherSeparatorNorValue()
        {
            var vars = new CalamariVariables
            {
                { "key6", "  new value" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void ShouldIgnoreOctopusPrefix()
        {
            var vars = new CalamariVariables
            {
                { "Octopus.Release.Id", "999" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void HandlesVariablesThatReferenceOtherVariables()
        {
            var vars = new CalamariVariables
            {
                { "key1", "#{a}" },
                { "a", "#{b}" },
                { "b", "new-value" },
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void UnicodeCharsInValuesAreEscaped()
        {
            var vars = new CalamariVariables
            {
                { "key1", "Θctopus" }
            };
            RunTest(vars, "example.properties");
        }

        [Test]
        public void UnicodeCharsInKeysAreHandled()
        {
            var vars = new CalamariVariables
            {
                { "kḛy", "new-value" }
            };
            RunTest(vars, "unicode-key.properties");
        }

        [Test]
        public void ShouldPreserveEncodingUtf8DosBom()
        {
            var vars = new CalamariVariables();
            RunHexTest(vars, "enc-utf8-dos-bom.properties");
        }

        [Test]
        public void ShouldPreserveEncodingUtf8UnixNoBom()
        {
            var vars = new CalamariVariables();
            RunHexTest(vars, "enc-utf8-unix-nobom.properties");
        }

        [Test]
        public void ShouldPreserveEncodingUtf16DosBom()
        {
            var vars = new CalamariVariables();
            RunHexTest(vars, "enc-utf16-dos-bom.properties");
        }

        [Test]
        public void ShouldPreserveEncodingWindows1252DosNoBom()
        {
            var vars = new CalamariVariables();
            RunHexTest(vars, "enc-windows1252-dos-nobom.properties");
        }

        void RunTest(CalamariVariables vars,
                     string file,
                     [CallerMemberName]
                     string testName = null,
                     [CallerFilePath]
                     string filePath = null)
        {
            // ReSharper disable once ExplicitCallerInfoArgument
            this.Assent(Replace(vars, file), AssentConfiguration.Properties, testName, filePath);
        }

        void RunHexTest(CalamariVariables vars,
                     string file,
                     [CallerMemberName]
                     string testName = null,
                     [CallerFilePath]
                     string filePath = null)
        {
            // ReSharper disable once ExplicitCallerInfoArgument
            this.Assent(ReplaceToHex(vars, file), AssentConfiguration.Default, testName, filePath);
        }
    }
}