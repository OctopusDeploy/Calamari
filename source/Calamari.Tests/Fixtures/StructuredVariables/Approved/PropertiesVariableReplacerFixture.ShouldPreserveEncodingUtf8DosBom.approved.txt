EF BB BF 64 61 67 67 65   ???dagge
72 3A E2 80 A0 0D 0A 73   r:???^^s
75 6E 6E 69 65 73 3A F0   unnies:?
9F 98 8E 0D 0A 65 75 72   ???^^eur
6F E2 82 AC               o???